OKTA-946202 OAuth 2.0 token lifecycle concept for OCI (#5554)
okta/okta-developer-docs: packages/@okta/vuepress-site/docs/concepts/token-lifecycles/index.md

* **Remote** (token introspection): For opaque tokens or for an additional layer of security with JWTs, the resource server sends the token to the authorization server's introspection endpoint.
  * The [introspect](https://developer.okta.com/docs/api/openapi/okta-oauth/oauth/tag/CustomAS/#tag/CustomAS/operation/introspectCustomAS) endpoint returns information about the token, including whether it's active, its scopes, client ID, and expiration. This method adds network overhead but provides the most up-to-date status, including immediate revocation status.

See [Validate access token](/docs/guides/validate-access-tokens/) for the Okta implementation.

#### ID token validation

The client app validates the ID token locally. This involves:

* **Signature verification**: Use the authorization server's public key to verify the signature.
* **Expiration check**: Confirm that the token hasn’t expired.
* **Issuer check**: Verify that the expected authorization server issued the token.
* **Audience check**: Ensure that the token is intended for this specific client app.

> * **Nonce check**: (If applicable) Verify that the nonce value to prevent replay attacks.

Reviewer: ```suggestion
* **Nonce check**: (If applicable) Verify that the nonce value exists. This value helps prevent replay attacks.
```

1. Authorization server validates and issues token: The authorization server validates the initial access token and grants a new access token to the initial API service. The new access token is limited to pre-approved scopes between the initial API service and the downstream API service and preserves the original user's context.
1. Downstream API service access with context: The API service uses this new token to access the downstream service, which can now understand on whose behalf the request is being made.

See [Set up OAuth 2.0 On-Behalf-of-Token-Exchange](/docs/guides/set-up-token-exchange/main/).

### Expiration

Access tokens and ID tokens are short-lived. The following are the results of an expired token:

* During access token validation, the resource server detects that the token's `exp` (expiration) claim indicates it's no longer valid.
* The client app, during ID token validation, detects that the ID token has expired.
* The resource server rejects the request with an HTTP 401 Unauthorized status code, indicating that the access token isn't valid for accessing protected resources.

> When an access token expires, the client app can't use it to access protected resources. To maintain continuous access without requiring the user to reauthenticate, the client app relies on the refresh token through the token renewal process. See [Renewal (refreshing tokens)](#renewal-refresh-tokens).

Reviewer: ```suggestion
When an access token expires, the client app can't use it to access protected resources. To maintain continuous access without requiring the user to reauthenticate, the client app relies on the refresh token through the token renewal process. See [Renewal (refresh tokens)](#renewal-refresh-tokens).
```

### Renewal (refresh tokens)

When an access token expires or is about to expire, the client app can use the refresh token to obtain a new access token (and often a new refresh token and ID token) from the authorization server. The following is the token renewal process:

1. The client app sends the refresh token to the authorization server's token endpoint with a `grant_type` of [`refresh_token`](https://developer.okta.com/docs/api/openapi/okta-oauth/oauth/tag/CustomAS/#tag/CustomAS/operation/tokenCustomAS!path=3/grant_type&t=request).
1. The authorization server validates the refresh token, checking its validity, expiration, and whether it has been revoked.
1. If the refresh token is valid, the authorization server issues the client a new access token (and potentially a new refresh token and ID token).
1. The client can then resume accessing protected resources using the newly acquired access token and verify the user's identity with the new ID token.

> This mechanism significantly improves user experience by minimizing the need for repeated logins, as the refresh token acts as a long-lived credential for acquiring short-lived access tokens. Public clients, such as browser-based apps, have a higher risk of compromised long-lived refresh token. You can reduce this risk by reducing the [refresh token lifetime](/docs/guides/refresh-tokens/main/#refresh-token-lifetime) and rotating refresh tokens after each use. See [Refresh access tokens and rotate refresh tokens](/docs/guides/refresh-tokens/main/).

Reviewer: ```suggestion
This mechanism significantly improves user experience by minimizing the need for the user to sign in again, as the refresh token acts as a long-lived credential for acquiring short-lived access tokens. Public clients, such as browser-based apps, have a higher risk of compromised long-lived refresh tokens. You can reduce this risk by reducing the [refresh token lifetime](/docs/guides/refresh-tokens/main/#refresh-token-lifetime) and rotating refresh tokens after each use. See [Refresh access tokens and rotate refresh tokens](/docs/guides/refresh-tokens/main/).
```

### Revocation (invalidating tokens)

Token revocation allows for the immediate invalidation of an access token or a refresh token before its natural expiration. This operation is required for security and lifecycle management.

**Purpose**: To instantly disable a token for use. Common scenarios include:

> * The user logs out of the app, requiring all associated tokens to be invalidated.

Reviewer: ```suggestion
* The user signs out of the app, requiring all associated tokens to be invalidated.
```

* The user changes their password, which might trigger the revocation of existing tokens for security.
* A security breach or compromise of a token is detected, needing its immediate invalidation.
* Administrator action to manually revoke a token.

**Mechanism**: The client app or an admin sends a request to the authorization server's [revocation](https://developer.okta.com/docs/api/openapi/okta-oauth/oauth/tag/CustomAS/#tag/CustomAS/operation/revokeCustomAS) endpoint, explicitly revoking the access or refresh token.

> **Effect**: Once revoked, the token is added to a blocklist or denylist maintained by the authorization server:

Reviewer: ```suggestion
**Effect**: After the token is revoked, it's added to a blocklist or denylist maintained by the authorization server:
```

See [Set up OAuth 2.0 On-Behalf-of-Token-Exchange](/docs/guides/set-up-token-exchange/main/).

### Expiration

Access tokens and ID tokens are short-lived. The following are the results of an expired token:

* During access token validation, the resource server detects that the token's `exp` (expiration) claim indicates it's no longer valid.
* The client app, during ID token validation, detects that the ID token has expired.
* The resource server rejects the request with an HTTP 401 Unauthorized status code, indicating that the access token isn't valid for accessing protected resources.

When an access token expires, the client app can't use it to access protected resources. To maintain continuous access without requiring the user to reauthenticate, the client app relies on the refresh token through the token renewal process. See [Renewal (refreshing tokens)](#renewal-refresh-tokens).

### Renewal (refresh tokens)

When an access token expires or is about to expire, the client app can use the refresh token to obtain a new access token (and often a new refresh token and ID token) from the authorization server. The following is the token renewal process:

> 1. The client app sends the refresh token to the authorization server's token endpoint with a `grant_type` of [`refresh_token`](https://developer.okta.com/docs/api/openapi/okta-oauth/oauth/tag/CustomAS/#tag/CustomAS/operation/tokenCustomAS!path=3/grant_type&t=request). This is typically a back-channel request that occurs directly between the client app and the authorization server, not through the user's browser.

Author: Please see my updates for this section (added some public app notes after the list).
